aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -524,6 +597,9 @@ SparseTensor& mul_out_sparse_cpu(SparseTensor& r, const Tensor& t_, const Tensor
     );
   }
 
+  r.resize_as_(src);
+  Tensor r_values = r_buffer.to(r.scalar_type());
```

Comment:
Could we just construct `r_values` to be the correct dtype and not do a `.to(...)` here?

Reply:
I think you're looking at another one of those cases where for ops like `float_tensor *= double_tensor` we need r_buffer.scalar_type()==double to be converted to the result type r_values.scalar_type()==float. We perform ops at the common/higher type but in some cases allow downcasting to a narrower output type (according to the rules of torch.can_cast).